File: sdks/python/engram_client/resources/resources.py
"""MCP Resource discovery, reading, and dynamic subscription resource mixin."""

from __future__ import annotations

from typing import Any

from .base import ResourceMixin


class McpResourcesMixin(ResourceMixin):
    """MCP Resource discovery and dynamic live subscription operations."""

    async def resource_list(self) -> dict[str, Any]:
        """List all resource templates exposed by the MCP server."""
        return await self._mcp_call("resources/list")

    async def resource_read(self, uri: str) -> dict[str, Any]:
        """Read an MCP resource by URI.

        Args:
            uri: The resource URI (e.g. `engram://stats`, `engram://memory/1`, `engram://workspace/dev/memories`).
        """
        return await self._mcp_call("resources/read", {"uri": uri})

    async def resource_subscribe(self, uri: str) -> dict[str, Any]:
        """Subscribe to live updates for an MCP resource URI.

        Args:
            uri: The resource URI to subscribe to.
        """
        return await self._mcp_call("resources/subscribe", {"uri": uri})

    async def resource_unsubscribe(self, uri: str) -> dict[str, Any]:
        """Unsubscribe from updates for an MCP resource URI.

        Args:
            uri: The resource URI to unsubscribe from.
        """
        return await self._mcp_call("resources/unsubscribe", {"uri": uri})

File: sdks/typescript/src/resources/resources.ts
import { BaseResource } from "./base.js";

/**
 * MCP Resource discovery, reading, and dynamic subscription management.
 */
export class McpResourcesResource extends BaseResource {
  /**
   * List all available resource URI templates from the server.
   */
  async list(): Promise<unknown> {
    return this.caller.mcpCall("resources/list");
  }

  /**
   * Read a resource by its URI.
   *
   * @param uri The resource URI (e.g. `engram://stats`, `engram://memory/1`, `engram://workspace/dev/memories`).
   */
  async read(uri: string): Promise<unknown> {
    return this.caller.mcpCall("resources/read", { uri });
  }

  /**
   * Subscribe to live updates for a resource URI.
   *
   * @param uri The resource URI to subscribe to.
   */
  async subscribe(uri: string): Promise<unknown> {
    return this.caller.mcpCall("resources/subscribe", { uri });
  }

  /**
   * Unsubscribe from updates for a resource URI.
   *
   * @param uri The resource URI to unsubscribe from.
   */
  async unsubscribe(uri: string): Promise<unknown> {
    return this.caller.mcpCall("resources/unsubscribe", { uri });
  }
}
